src/routes/auth.rs

```diff
@@ -75,27 +86,66 @@ impl<'a, 'r> FromRequest<'a, 'r> for User {
         };
 
-        // Check if a key is required for authentication
-        if !auth_data.key_required() {
-            return Outcome::Success(User::create_and_store_user(request, &auth_data));
-        }
+        let key_opt = request.headers().get_one(AUTH_HEADER);
+        let env: State<Env> = match request.guard().succeeded() {
+            Some(env) => env,
+            None => return Error::from(ErrorKind::Unknown).into_outcome()
+        };
+        let ldap_config = &env.config().ldap;
 
-        // Check the user's key, if provided
-        if let Some(key) = request.headers().get_one(AUTH_HEADER) {
-            if auth_data.key_matches(key) {
-                // The key matches, so create and store a new user and cookie
-                Outcome::Success(Self::create_and_store_user(request, &auth_data))
-            } else {
-                // The key does not match
-                Error::from(ErrorKind::Unauthorized).into_outcome()
+        if ldap_config.enabled {
+            println!("LDAP is enabled");
+            let key = match key_opt {
+                Some(key) => key,
+                None => return Error::from(ErrorKind::LdapMissingKey).into_outcome()
+            };
+            let username = match request.headers().get_one(AUTH_HEADER_USERNAME) {
+                Some(username) => username,
+                None => return Error::from(ErrorKind::LdapMissingUsername).into_outcome()
+            };
+
+            match ldap_login(&ldap_config.address, &ldap_config.bind_dn, username, key) {
+                Ok(_) => Outcome::Success(User::create_and_store_user(request, &auth_data)),
+                Err(e) => e.into_outcome()
             }
+
         } else {
+            
+            // Check if a key is required for authentication
+            if !auth_data.key_required() {
+                return Outcome::Success(User::create_and_store_user(request, &auth_data));
+            }
+
+            // Check the user's key, if provided
+            if let Some(key) = key_opt {
+                if auth_data.key_matches(key) {
+                    // The key matches, so create and store a new user and cookie
+                    Outcome::Success(Self::create_and_store_user(request, &auth_data))
+                } else {
+                    // The key does not match
+                    Error::from(ErrorKind::Unauthorized).into_outcome()
+                }
+            } else {
             // A key is required but not provided
-            Error::from(ErrorKind::Unauthorized).into_outcome()
+                Error::from(ErrorKind::Unauthorized).into_outcome()
+            } 
+
         }
     }
 }
 
+fn ldap_login(ldap_address: &str, bind_dn: &str, username: &str, key: &str) -> Result<(), Error> {
+    println!("LDAP address: {}, user: {}", ldap_address, username);
+    let bind_dn = bind_dn.replace("{}", username);
+    LdapConn::new(&ldap_address)
+        .map_err(|e| ErrorKind::LdapConnectError(format!("{:?}", e)))?
```

Comment:
This should use `.context()` to wrap the error with context. Some error details may be lost in the conversion to string (ex. causes). The exact error details will then be logged server-side. The user of the API does not need to know the underlying error, as that is the business of whoever is running the API. The ErrorKind then should not have any associated string value.

If there are some errors which should be shown to the user, please share them and it may make sense to include them in the API's output. However, it looks like the errors would just be general IO errors (the error type is `std::io::Error`).